### PR TITLE
Add length[] function

### PR DESCRIPTION
**Pull request by cmontella**

### CODE REVIEW

**Comment by cmontella** [comment]
An extension of https://github.com/witheve/Eve/pull/644

Credit goes to @jimmyhmiller for this one!

**Comment by cmontella** [comment]
A follow-on to the PR is the documentation for length[]: https://github.com/witheve/docs/pull/60

File: src/runtime/providers/string.ts [hunk 1]
```diff
@@ -251,7 +251,83 @@ class Urlencode extends Constraint {
...
+  // Adapted from: https://mathiasbynens.be/notes/javascript-unicode
+  countSymbols(string) {
+    let index;
+    let symbolCount = 0;
+    for (index = 0; index < string.length - 1; ++index) {
+      var charCode = string.charCodeAt(index);
+      if (charCode >= 0xD800 && charCode <= 0xDBFF) {
+        charCode = string.charCodeAt(index + 1);
+        if (charCode >= 0xDC00 && charCode <= 0xDFFF) {
+          ++index;
```
Comment (joshuafcole) [inline src/runtime/providers/string.ts]: Super unimportant, but let's avoid using the pre-increment for clarity reasons.

File: src/runtime/providers/string.ts [hunk 2]
```diff
@@ -251,7 +251,83 @@ class Urlencode extends Constraint {
...
+    } else if (az === "code-points") {
+      return [text.length];
+    }
+    return undefined;
+  }
+
+  resolveProposal(proposal, prefix) {
+    let {args} = this.resolve(prefix);
+    let [text, az] = args;
+    if (az === undefined) {az = "symbols"};
```
Comment (joshuafcole) [inline src/runtime/providers/string.ts]: Let's either drop the braces or indent this appropriately. This looks confusingly like an object.

File: test/strings.ts [hunk 3]
```diff
@@ -25,3 +25,154 @@ test("test string join ordering", (assert) => {
   `);
   assert.end();
 });
+
+test("check length of hello", (assert) => {
```
Comment (joshuafcole) [inline test/strings.ts]: These may need touched depending on whether they go in before the async meat-grinder work Chris has going on.

File: test/strings.ts [hunk 4]
```diff
@@ -25,3 +25,154 @@ test("test string join ordering", (assert) => {
...
+      len = length[text: "hello" as: "asdfasdf"]
+
+    commit
+      [len: len]
+    ~~~
+  `);
+  assert.end();
+})
+
+test("check length multi-byte characters as symbols", (assert) => {
```
Comment (joshuafcole) [inline test/strings.ts]: We should add a test-case where symbols and code-points would differ.

File: test/strings.ts (hunk 3 above)
Comment (cmontella) [inline test/strings.ts]: What do you mean by "touched" ?

**Comment by joshuafcole** [comment]
✈️  🛩  ☁️

**Pull request merged**